### ISSUE TITLE
Topics which has conkeyref should be processed dynamically

### ISSUE THREAD

**Issue by ToshihikoMakita** [dita-ot/dita-ot#2525]
I believe that DITA 1.3 allows duplicate key name definitions if they are defined in the different key scopes.

For instance the following two tasks tWhenConnectingASinglePrinter.dita and tWhenConnectingMultiplePrinters.dita has same conkeyref  authoring ``<stepsection conkeyref="CONNECTING_PRINTERS/stepsection_01"/>`` that references the following part of tWhenConnectingPrintersTemplate.dita .

```xml
<stepsection id="stepsection_01">To configure the settings of the color measurement instrument, 
go to step <xref keyref="REFERENCE-TOPIC/step-color" format="dita" type="step"/>. 
Otherwise, go to step <xref keyref="REFERENCE-TOPIC/step-other" format="dita" type="step"/>.</stepsection>
```

The base bookmap has followoing structure:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE bookmap PUBLIC "-//OASIS//DTD DITA BookMap//EN" "bookmap.dtd">
<bookmap xml:lang="en-US">
    <booktitle>
        <mainbooktitle>Online Help</mainbooktitle>
    </booktitle>
    <frontmatter>
        <mapref href="mKeyDef.ditamap"/>
    </frontmatter>
    <chapter href="topics/cUsingOnlineHelp.dita" navtitle="How to use online-help"/>
    <chapter navtitle="Prepareing to use the software" href="topics/cPreparingToUseSoftware.dita">
        <topicgroup keyscope="SINGLE-PRINTER">
            <keydef keys="CONNECTING_PRINTERS" href="collection-topics/tWhenConnectingPrintersTemplate.dita"/>
            <topicref navtitle="Connecting a single printer" keys="REFERENCE-TOPIC" format="dita" href="topics/tWhenConnectingASinglePrinter.dita"/>
        </topicgroup>
        <topicgroup keyscope="MULTIPLE-PRINTER">
            <keydef keys="CONNECTING_PRINTERS" href="collection-topics/tWhenConnectingPrintersTemplate.dita"/>
            <topicref navtitle="Connecting multiple printer"  keys="REFERENCE-TOPIC" format="dita" href="topics/tWhenConnectingMultiplePrinters.dita"/>
        </topicgroup>
    </chapter>
</bookmap>
```

However in the middle file in the temp directory, tWhenConnectingPrintersTemplate.dita is expanded as follows:

[temp/collection-topics/tWhenConnectingPrintersTemplate.dita]

```xml
<stepsection id="stepsection_01" class="- topic/li task/stepsection ">
To configure the settings of the color measurement instrument, go to step 
<xref keyref="REFERENCE-TOPIC/step-color" format="dita" type="step" class="- topic/xref " 
href="../topics/tWhenConnectingMultiplePrinters.dita#task_ops_1q4_wx/step-color">
<?ditaot usertext?>Connecting a single printer</xref>. 
Otherwise, go to step <xref keyref="REFERENCE-TOPIC/step-other" format="dita" type="step" 
class="- topic/xref " href="../topics/tWhenConnectingMultiplePrinters.dita#task_ops_1q4_wx/step-other">
<?ditaot usertext?>Connecting a single printer</xref>.</stepsection>
```
As you see above the collection topic (reused topic) is processed statically using *FIXED* "REFERENCE-TOPIC" key value (topics/tWhenConnectingMultiplePrinters.dita). So multiple "REFERENCE-TOPIC" keydefs in the different key scope does not have any meaning.

As the result XHTML output has irregular xref/@href value and xref/text().

[tWhenConnectingASinglePrinter.html]

```html
<p class="li stepsection">To configure the settings of the color measurement instrument, go to step
<a class="xref" href="../collection-topics/../topics/tWhenConnectingMultiplePrinters.html#task_ops_1q4_wx__step-color">
Connecting a single printer</a>. 
Otherwise, go to step <a class="xref" href="../collection-topics/../topics/tWhenConnectingMultiplePrinters.html#task_ops_1q4_wx__step-other">
Connecting a single printer</a>.</p>
```

[tWhenConnectingMultiplePrinters.html]
```html
<p class="li stepsection">To configure the settings of the color measurement instrument, go to step 
<a class="xref" href="../collection-topics/../topics/tWhenConnectingMultiplePrinters.html#task_ops_1q4_wx__step-color">
Connecting a single printer</a>. 
Otherwise, go to step <a class="xref" href="../collection-topics/../topics/tWhenConnectingMultiplePrinters.html#task_ops_1q4_wx__step-other">
Connecting a single printer</a>.</p>
```

I think this is DITA-OT serious design problem for scoped key processing. Please fix this problem.

I attached sample instance.

[20161119-key-scope.zip](https://github.com/dita-ot/dita-ot/files/600444/20161119-key-scope.zip)

Regards,

Toshihiko Makita

**Comment by raducoravu** [comment]
+1

One of our clients wanted to reuse a table row with slightly different content in two places.
Again, conkeyref does not seem to properly resolve to proper content depending on key scope.
I'm attaching my samples:

[conkeyrefKeyScopes.zip](https://github.com/dita-ot/dita-ot/files/770455/conkeyrefKeyScopes.zip)

**Comment by raducoravu** [comment]
👍 An extra report which seems to be about the same problem (and it can be reproduced with DITA OT 3.x):
https://www.oxygenxml.com/forum/post51846.html#p51846

Attaching the test file to the issue:
[DITA Test.zip](https://github.com/dita-ot/dita-ot/files/2831613/DITA.Test.zip)

**Comment by raducoravu** [comment]
Problem can still be reproduced with DITA OT 3.4

**Comment by chrispy-snps** [comment]
It looks like this was fixed in DITA-OT 3.7.

From Makita-san's original testcase:

```
$ diff -r out-3.6.1 out-3.7
diff -r out-3.6.1/topics/tWhenConnectingMultiplePrinters.html out-3.7/topics/tWhenConnectingMultiplePrinters.html
39c39
< </ol><p class="li stepsection">To configure the settings of the color measurement instrument, go to step <a class="xref" href="../collection-topics/../topics/tWhenConnectingASinglePrinter.html#task_lbb_mc4_wx__step-color">3</a>. Otherwise, go to step <a class="xref" href="../collection-topics/../topics/tWhenConnectingASinglePrinter.html#task_lbb_mc4_wx__step-other">4</a>.</p><ol start="4"><li class="li step" id="task_w4d_yjq_wx__step-color">
---
> </ol><p class="li stepsection">To configure the settings of the color measurement instrument, go to step <a class="xref" href="../collection-topics/../topics/tWhenConnectingMultiplePrinters.html#task_ops_1q4_wx__step-color">4</a>. Otherwise, go to step <a class="xref" href="../collection-topics/../topics/tWhenConnectingMultiplePrinters.html#task_ops_1q4_wx__step-other">5</a>.</p><ol start="4"><li class="li step" id="task_w4d_yjq_wx__step-color">
```

From the conkeyrefKeyScopes.zip testcase:

```
$ diff -r out-3.6.1 out-3.7
diff -r out-3.6.1/t1.html out-3.7/t1.html
...
14c14
<                         <td class="entry">value <span class="ph">SECOND PH Content</span></td>
---
>                         <td class="entry">value <span class="ph">First PH Content</span></td>
```

From the DITA Test.zip testcase:

```
$ diff -r out-3.6.1 out-3.7
diff -r out-3.6.1/Chapter1/Topic1.html out-3.7/Chapter1/Topic1.html
24c24
<                     <p class="p">This Template text refers to the special text of: <a class="xref" href="../TemplatesFolder/../Chapter2/Topic2.html">Topic 2</a></p>
---
>                     <p class="p">This Template text refers to the special text of: <a class="xref" href="../TemplatesFolder/../Chapter1/Topic1.html">Topic 1</a></p>
```

**Comment by raducoravu** [comment]
@chrispy-snps thanks for checking this, I'll close this, possibly it was fixed as a side effect of another issue.

**Comment by chrispy-snps** [comment]
@raducoravu - thanks for closing this. I was going to try to narrow the fix down to a specific commit between 3.6.1 and 3.7, but I got some kind of certificate error from `gradlew` when I attempted to build 3.6.1 as a baseline, and I was too lazy to dig into it. :)
